File: src/sdk/python/amc_client.py
"""
AMC Python SDK — Agent Maturity Compass

A Python client for the AMC Bridge HTTP API. Provides typed access to all
provider endpoints (OpenAI, Anthropic, Gemini, OpenRouter, xAI, local),
telemetry reporting, evidence hashing, and secret redaction.

Usage:
    from amc_client import AMCClient

    client = AMCClient(bridge_url="http://localhost:3212", token="your-token")
    response = client.openai_chat({"model": "gpt-4", "messages": [{"role": "user", "content": "Hello"}]})
    print(response.body)

Environment variables:
    AMC_BRIDGE_URL: Bridge server URL (default: http://localhost:3212)
    AMC_TOKEN: Authentication token
    AMC_WORKSPACE_ID: Optional workspace identifier
"""

from __future__ import annotations

import hashlib
import json
import os
import re
import uuid
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Union


# ---------------------------------------------------------------------------
# Types
# ---------------------------------------------------------------------------

@dataclass
class AMCBridgeResponse:
    """Wrapper for Bridge API responses."""
    status: int
    body: Any
    request_id: Optional[str] = None
    receipt: Optional[str] = None
    correlation_id: Optional[str] = None
    deprecated: bool = False
    sunset: Optional[str] = None
    warning: Optional[str] = None

    @property
    def ok(self) -> bool:
        return 200 <= self.status < 300


@dataclass
class AMCClientConfig:
    """Configuration for the AMC client."""
    bridge_url: str = ""
    token: str = ""
    workspace_id: Optional[str] = None
    timeout: float = 30.0
    verify_ssl: bool = True


# ---------------------------------------------------------------------------
# Redaction & Hashing utilities
# ---------------------------------------------------------------------------

_SECRET_PATTERNS: List[re.Pattern] = [
    re.compile(r"sk-[A-Za-z0-9]{10,}"),
    re.compile(r"(?:api|secret|token|key)\s*[:=]\s*[A-Za-z0-9._-]{10,}", re.IGNORECASE),
    re.compile(r"bearer\s+[A-Za-z0-9._-]{10,}", re.IGNORECASE),
    re.compile(r"-----BEGIN [A-Z ]*PRIVATE KEY-----"),
]


def redact_text(text: str) -> str:
    """Remove secret-like patterns from text."""
    result = text
    for pattern in _SECRET_PATTERNS:
        result = pattern.sub("[REDACTED]", result)
    return result


def hash_value(value: Union[str, dict, list]) -> str:
    """SHA256 hash of a value (string or JSON-serializable object)."""
    if isinstance(value, str):
        data = value
    else:
        data = json.dumps(value, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(data.encode("utf-8")).hexdigest()


def _assert_no_self_scoring(payload: dict) -> None:
    """Prevent the SDK from scoring its own outputs."""
    messages = payload.get("messages", [])
    for msg in messages:
        content = msg.get("content", "")
        if isinstance(content, str) and "amc_self_score" in content.lower():
            raise ValueError(
                "Self-scoring detected: AMC SDK cannot be used to score its own outputs."
            )


# ---------------------------------------------------------------------------
# AMCClient
# ---------------------------------------------------------------------------

class AMCClient:
    """
    Python client for the AMC Bridge HTTP API.

    Supports all provider endpoints (OpenAI, Anthropic, Gemini, OpenRouter,
    xAI, local), telemetry, and evidence utilities.

    Args:
        bridge_url: Bridge server URL. Falls back to AMC_BRIDGE_URL env var.
        token: Authentication token. Falls back to AMC_TOKEN env var.
        workspace_id: Optional workspace identifier.
        timeout: HTTP request timeout in seconds (default: 30).
        verify_ssl: Whether to verify SSL certificates (default: True).
    """

    def __init__(
        self,
        bridge_url: Optional[str] = None,
        token: Optional[str] = None,
        workspace_id: Optional[str] = None,
        timeout: float = 30.0,
        verify_ssl: bool = True,
    ):
        self.bridge_url = (bridge_url or os.getenv("AMC_BRIDGE_URL", "http://localhost:3212")).rstrip("/")
        self.token = token or os.getenv("AMC_TOKEN", "")
        self.workspace_id = workspace_id or os.getenv("AMC_WORKSPACE_ID")
        self.timeout = timeout
        self.verify_ssl = verify_ssl

        if not self.bridge_url:
            raise ValueError(
                "bridge_url is required. Set AMC_BRIDGE_URL env var or pass bridge_url parameter."
            )

    @classmethod
    def from_env(cls, **kwargs) -> "AMCClient":
        """Create a client using AMC_* environment variables with optional overrides."""
        return cls(**kwargs)

    def _headers(self, correlation_id: Optional[str] = None, extra: Optional[Dict[str, str]] = None) -> Dict[str, str]:
        """Build request headers."""
        headers: Dict[str, str] = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {self.token}",
            "x-amc-correlation-id": correlation_id or str(uuid.uuid4()),
            "x-amc-sdk-name": "amc-python-sdk",
            "x-amc-sdk-version": "0.1.0",
        }
        if self.workspace_id:
            headers["x-amc-workspace-id"] = self.workspace_id
        if extra:
            headers.update(extra)
        return headers

    def _call_bridge(
        self,
        path: str,
        payload: dict,
        correlation_id: Optional[str] = None,
        extra_headers: Optional[Dict[str, str]] = None,
    ) -> AMCBridgeResponse:
        """
        Make a POST request to the Bridge API.

        Uses urllib.request (stdlib) to avoid external dependencies.
        For production use with async support, consider using httpx.
        """
        import urllib.request
        import urllib.error
        import urllib.parse

        url = f"{self.bridge_url}{path}"
        headers = self._headers(correlation_id, extra_headers)
        data = json.dumps(payload).encode("utf-8")

        req = urllib.request.Request(url, data=data, headers=headers, method="POST")

        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                body_bytes = resp.read()
                try:
                    body = json.loads(body_bytes)
                except (json.JSONDecodeError, ValueError):
                    body = body_bytes.decode("utf-8", errors="replace")

                return AMCBridgeResponse(
                    status=resp.status,
                    body=body,
                    request_id=resp.headers.get("x-amc-bridge-request-id"),
                    receipt=resp.headers.get("x-amc-receipt"),
                    correlation_id=resp.headers.get("x-amc-correlation-id"),
                    deprecated=(resp.headers.get("deprecation", "").lower() == "true"),
                    sunset=resp.headers.get("sunset"),
                    warning=resp.headers.get("warning"),
                )
        except urllib.error.HTTPError as e:
            body_bytes = e.read()
            try:
                body = json.loads(body_bytes)
            except (json.JSONDecodeError, ValueError):
                body = body_bytes.decode("utf-8", errors="replace")

            return AMCBridgeResponse(
                status=e.code,
                body=body,
                request_id=e.headers.get("x-amc-bridge-request-id") if e.headers else None,
                receipt=e.headers.get("x-amc-receipt") if e.headers else None,
                correlation_id=e.headers.get("x-amc-correlation-id") if e.headers else None,
                deprecated=((e.headers.get("deprecation", "").lower() == "true") if e.headers else False),
                sunset=e.headers.get("sunset") if e.headers else None,
                warning=e.headers.get("warning") if e.headers else None,
            )

    # ── Provider methods ──────────────────────────────────────────────

    def openai_chat(self, payload: dict, **kwargs) -> AMCBridgeResponse:
        """Send a chat completion request via OpenAI provider."""
        _assert_no_self_scoring(payload)
        return self._call_bridge("/bridge/openai/v1/chat/completions", payload, **kwargs)

    def openai_responses(self, payload: dict, **kwargs) -> AMCBridgeResponse:
        """Send a request to the OpenAI responses endpoint."""
        _assert_no_self_scoring(payload)
        return self._call_bridge("/bridge/openai/v1/responses", payload, **kwargs)

    def openai_batches(self, payload: dict, **kwargs) -> AMCBridgeResponse:
        """Send a request to the OpenAI batches endpoint."""
        _assert_no_self_scoring(payload)
        return self._call_bridge("/bridge/openai/v1/batches", payload, **kwargs)

    def anthropic_messages(self, payload: dict, **kwargs) -> AMCBridgeResponse:
        """Send a messages request via Anthropic provider."""
        _assert_no_self_scoring(payload)
        return self._call_bridge("/bridge/anthropic/v1/messages", payload, **kwargs)

    def gemini_generate_content(self, model: str, payload: dict, **kwargs) -> AMCBridgeResponse:
        """Send a generateContent request via Gemini provider."""
        _assert_no_self_scoring(payload)
        import urllib.parse
        model_path = urllib.parse.quote(model, safe="")
        return self._call_bridge(
            f"/bridge/gemini/v1beta/models/{model_path}:generateContent", payload, **kwargs
        )

    def openrouter_chat(self, payload: dict, **kwargs) -> AMCBridgeResponse:
        """Send a chat completion request via OpenRouter provider."""
        _assert_no_self_scoring(payload)
        return self._call_bridge("/bridge/openrouter/v1/chat/completions", payload, **kwargs)

    def xai_chat(self, payload: dict, **kwargs) -> AMCBridgeResponse:
        """Send a chat completion request via xAI/Grok provider."""
        _assert_no_self_scoring(payload)
        return self._call_bridge("/bridge/xai/v1/chat/completions", payload, **kwargs)

    def local_chat(self, payload: dict, **kwargs) -> AMCBridgeResponse:
        """Send a chat completion request via local (OpenAI-compatible) provider."""
        _assert_no_self_scoring(payload)
        return self._call_bridge("/bridge/local/v1/chat/completions", payload, **kwargs)

    # ── Telemetry ─────────────────────────────────────────────────────

    def report_telemetry(
        self,
        session_id: str,
        event_type: str,
        payload: Union[str, dict],
        *,
        correlation_id: Optional[str] = None,
        run_id: Optional[str] = None,
        provider: Optional[str] = None,
    ) -> AMCBridgeResponse:
        """Send a telemetry event to the Bridge."""
        body: Dict[str, Any] = {
            "sessionId": session_id,
            "eventType": event_type,
            "payload": redact_text(payload) if isinstance(payload, str) else payload,
        }
        if correlation_id:
            body["correlationId"] = correlation_id
        if run_id:
            body["runId"] = run_id
        if provider:
            body["provider"] = provider

        return self._call_bridge("/bridge/telemetry", body)

    def report_output(
        self,
        session_id: str,
        value: Union[str, dict],
        *,
        provider: Optional[str] = None,
        run_id: Optional[str] = None,
    ) -> AMCBridgeResponse:
        """Report agent output as a telemetry event."""
        if isinstance(value, str):
            payload_value = redact_text(value)
        else:
            payload_value = json.loads(redact_text(json.dumps(value)))
        return self.report_telemetry(
            session_id=session_id,
            event_type="agent_stdout",
            payload=payload_value,
            run_id=run_id,
            provider=provider,
        )

    # ── Evidence utilities ────────────────────────────────────────────

    @staticmethod
    def output_hash(value: Union[str, dict, list]) -> str:
        """Compute SHA256 hash of output for evidence verification."""
        return hash_value(value)

    @staticmethod
    def redact(text: str) -> str:
        """Redact secret-like patterns from text."""
        return redact_text(text)


# ---------------------------------------------------------------------------
# Factory
# ---------------------------------------------------------------------------

def create_amc_client(
    bridge_url: Optional[str] = None,
    token: Optional[str] = None,
    workspace_id: Optional[str] = None,
    **kwargs,
) -> AMCClient:
    """Create an AMCClient instance with optional configuration."""
    return AMCClient(
        bridge_url=bridge_url,
        token=token,
        workspace_id=workspace_id,
        **kwargs,
    )

File: src/sdk/go/amc_client.go
// Package amc provides a Go client for the Agent Maturity Compass Bridge HTTP API.
//
// It supports all provider endpoints (OpenAI, Anthropic, Gemini, OpenRouter, xAI, local),
// telemetry reporting, evidence hashing, lease management, and secret redaction.
//
// Usage:
//
//	client := amc.NewClient(amc.Config{BridgeURL: "http://localhost:3212", Token: "your-token"})
//	resp, err := client.OpenAIChat(ctx, map[string]any{"model": "gpt-4", "messages": []any{...}})
package amc

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"regexp"
	"strings"
	"time"

	"github.com/google/uuid"
)

// Config holds configuration for the AMC client.
type Config struct {
	BridgeURL   string
	Token       string
	WorkspaceID string
	Timeout     time.Duration
}

// BridgeResponse wraps an HTTP response from the Bridge API.
type BridgeResponse struct {
	Status        int
	Body          json.RawMessage
	RequestID     string
	Receipt       string
	CorrelationID string
}

// OK returns true if the response status is 2xx.
func (r *BridgeResponse) OK() bool {
	return r.Status >= 200 && r.Status < 300
}

// Decode unmarshals the body into the given target.
func (r *BridgeResponse) Decode(target any) error {
	return json.Unmarshal(r.Body, target)
}

// Client is the AMC Bridge HTTP client.
type Client struct {
	cfg  Config
	http *http.Client
}

// NewClient creates a new AMC client. It reads AMC_BRIDGE_URL, AMC_TOKEN,
// and AMC_WORKSPACE_ID from the environment as defaults.
func NewClient(cfg Config) *Client {
	if cfg.BridgeURL == "" {
		cfg.BridgeURL = os.Getenv("AMC_BRIDGE_URL")
	}
	if cfg.BridgeURL == "" {
		cfg.BridgeURL = "http://localhost:3212"
	}
	cfg.BridgeURL = strings.TrimRight(cfg.BridgeURL, "/")

	if cfg.Token == "" {
		cfg.Token = os.Getenv("AMC_TOKEN")
	}
	if cfg.WorkspaceID == "" {
		cfg.WorkspaceID = os.Getenv("AMC_WORKSPACE_ID")
	}
	if cfg.Timeout == 0 {
		cfg.Timeout = 30 * time.Second
	}

	return &Client{
		cfg:  cfg,
		http: &http.Client{Timeout: cfg.Timeout},
	}
}

// NewClientFromEnv creates a client entirely from AMC_* env vars.
func NewClientFromEnv() *Client {
	return NewClient(Config{})
}

func (c *Client) callBridge(ctx context.Context, path string, payload any, correlationID string) (*BridgeResponse, error) {
	data, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("amc: marshal payload: %w", err)
	}

	url := c.cfg.BridgeURL + path
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(data))
	if err != nil {
		return nil, fmt.Errorf("amc: create request: %w", err)
	}

	if correlationID == "" {
		correlationID = uuid.New().String()
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+c.cfg.Token)
	req.Header.Set("X-Amc-Correlation-Id", correlationID)
	if c.cfg.WorkspaceID != "" {
		req.Header.Set("X-Amc-Workspace-Id", c.cfg.WorkspaceID)
	}

	resp, err := c.http.Do(req)
	if err != nil {
		return nil, fmt.Errorf("amc: request failed: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("amc: read response: %w", err)
	}

	return &BridgeResponse{
		Status:        resp.StatusCode,
		Body:          json.RawMessage(body),
		RequestID:     resp.Header.Get("X-Amc-Bridge-Request-Id"),
		Receipt:       resp.Header.Get("X-Amc-Receipt"),
		CorrelationID: resp.Header.Get("X-Amc-Correlation-Id"),
	}, nil
}

// --- Provider methods ---

// OpenAIChat sends a chat completion request via the OpenAI provider.
func (c *Client) OpenAIChat(ctx context.Context, payload map[string]any) (*BridgeResponse, error) {
	if err := assertNoSelfScoring(payload); err != nil {
		return nil, err
	}
	return c.callBridge(ctx, "/bridge/openai/v1/chat/completions", payload, "")
}

// OpenAIResponses sends a request to the OpenAI responses endpoint.
func (c *Client) OpenAIResponses(ctx context.Context, payload map[string]any) (*BridgeResponse, error) {
	if err := assertNoSelfScoring(payload); err != nil {
		return nil, err
	}
	return c.callBridge(ctx, "/bridge/openai/v1/responses", payload, "")
}

// AnthropicMessages sends a messages request via the Anthropic provider.
func (c *Client) AnthropicMessages(ctx context.Context, payload map[string]any) (*BridgeResponse, error) {
	if err := assertNoSelfScoring(payload); err != nil {
		return nil, err
	}
	return c.callBridge(ctx, "/bridge/anthropic/v1/messages", payload, "")
}

// GeminiGenerateContent sends a generateContent request via the Gemini provider.
func (c *Client) GeminiGenerateContent(ctx context.Context, model string, payload map[string]any) (*BridgeResponse, error) {
	if err := assertNoSelfScoring(payload); err != nil {
		return nil, err
	}
	return c.callBridge(ctx, fmt.Sprintf("/bridge/gemini/v1beta/models/%s:generateContent", url.PathEscape(model)), payload, "")
}

// OpenRouterChat sends a chat completion request via OpenRouter.
func (c *Client) OpenRouterChat(ctx context.Context, payload map[string]any) (*BridgeResponse, error) {
	if err := assertNoSelfScoring(payload); err != nil {
		return nil, err
	}
	return c.callBridge(ctx, "/bridge/openrouter/v1/chat/completions", payload, "")
}

// XAIChat sends a chat completion request via xAI/Grok.
func (c *Client) XAIChat(ctx context.Context, payload map[string]any) (*BridgeResponse, error) {
	if err := assertNoSelfScoring(payload); err != nil {
		return nil, err
	}
	return c.callBridge(ctx, "/bridge/xai/v1/chat/completions", payload, "")
}

// LocalChat sends a chat completion request via local (OpenAI-compatible) provider.
func (c *Client) LocalChat(ctx context.Context, payload map[string]any) (*BridgeResponse, error) {
	if err := assertNoSelfScoring(payload); err != nil {
		return nil, err
	}
	return c.callBridge(ctx, "/bridge/local/v1/chat/completions", payload, "")
}

// --- Telemetry ---

// TelemetryEvent represents a telemetry event payload.
type TelemetryEvent struct {
	SessionID     string
	EventType     string
	Payload       any
	CorrelationID string
	RunID         string
	Provider      string
}

// ReportTelemetry sends a telemetry event to the Bridge.
func (c *Client) ReportTelemetry(ctx context.Context, evt TelemetryEvent) (*BridgeResponse, error) {
	body := map[string]any{
		"sessionId": evt.SessionID,
		"eventType": evt.EventType,
		"payload":   evt.Payload,
	}
	if evt.CorrelationID != "" {
		body["correlationId"] = evt.CorrelationID
	}
	if evt.RunID != "" {
		body["runId"] = evt.RunID
	}
	if evt.Provider != "" {
		body["provider"] = evt.Provider
	}
	return c.callBridge(ctx, "/bridge/telemetry", body, evt.CorrelationID)
}

// ReportOutput reports agent output as a telemetry event.
func (c *Client) ReportOutput(ctx context.Context, sessionID string, value any) (*BridgeResponse, error) {
	return c.ReportTelemetry(ctx, TelemetryEvent{
		SessionID: sessionID,
		EventType: "agent_stdout",
		Payload:   value,
	})
}

// --- Evidence utilities ---

// OutputHash computes SHA256 hash of output for evidence verification.
func OutputHash(value any) string {
	var data []byte
	switch v := value.(type) {
	case string:
		data = []byte(v)
	default:
		data, _ = json.Marshal(v)
	}
	h := sha256.Sum256(data)
	return fmt.Sprintf("%x", h)
}

// --- Redaction ---

var secretPatterns = []*regexp.Regexp{
	regexp.MustCompile(`sk-[A-Za-z0-9]{10,}`),
	regexp.MustCompile(`(?i)(?:api|secret|token|key)\s*[:=]\s*[A-Za-z0-9._-]{10,}`),
	regexp.MustCompile(`(?i)bearer\s+[A-Za-z0-9._-]{10,}`),
	regexp.MustCompile(`-----BEGIN [A-Z ]*PRIVATE KEY-----`),
}

// Redact removes secret-like patterns from text.
func Redact(text string) string {
	result := text
	for _, pat := range secretPatterns {
		result = pat.ReplaceAllString(result, "[REDACTED]")
	}
	return result
}

// --- Self-scoring guard ---

func assertNoSelfScoring(payload map[string]any) error {
	messages, ok := payload["messages"]
	if !ok {
		return nil
	}
	msgSlice, ok := messages.([]any)
	if !ok {
		return nil
	}
	for _, msg := range msgSlice {
		m, ok := msg.(map[string]any)
		if !ok {
			continue
		}
		content, ok := m["content"].(string)
		if !ok {
			continue
		}
		if strings.Contains(strings.ToLower(content), "amc_self_score") {
			return errors.New("amc: self-scoring detected — AMC SDK cannot be used to score its own outputs")
		}
	}
	return nil
}

// --- Lease management ---

// LeaseInfo holds lease metadata.
type LeaseInfo struct {
	LeaseID   string   `json:"leaseId"`
	AgentID   string   `json:"agentId"`
	ExpiresAt string   `json:"expiresAt"`
	Scopes    []string `json:"scopes"`
}

// RequestLease requests a new lease from Studio lease routes.
func (c *Client) RequestLease(ctx context.Context, agentID string, scopes []string, durationSec int) (*BridgeResponse, error) {
	if durationSec <= 0 {
		durationSec = 3600
	}
	scopeList := strings.Join(scopes, ",")
	if scopeList == "" {
		scopeList = "gateway:llm,toolhub:intent,toolhub:execute"
	}
	body := map[string]any{
		"agentId": agentID,
		"ttl":     fmt.Sprintf("%ds", durationSec),
		"scopes":  scopeList,
	}
	return c.callBridge(ctx, "/leases/issue", body, "")
}

// RevokeLease revokes an active lease.
func (c *Client) RevokeLease(ctx context.Context, leaseID string) (*BridgeResponse, error) {
	body := map[string]any{"leaseId": leaseID}
	return c.callBridge(ctx, "/leases/revoke", body, "")
}

// --- Diagnostic ---

// QueryDiagnostic runs a diagnostic query for an agent.
func (c *Client) QueryDiagnostic(ctx context.Context, agentID string, questionIDs []string) (*BridgeResponse, error) {
	body := map[string]any{
		"agentId":     agentID,
		"questionIds": questionIDs,
	}
	return c.callBridge(ctx, "/bridge/diagnostic/query", body, "")
}
